src/shogun/kernel/PeriodicKernel.cpp

```diff
@@ -0,0 +1,169 @@
+		float64_t w, float64_t p, int32_t size) : CDotKernel(size)
+{
+	init();
+	set_width(w);
+	set_period(p);
+	init(l,r);
+}
+
+CPeriodicKernel::~CPeriodicKernel()
+{
+	cleanup();
+}
+
+CPeriodicKernel* CPeriodicKernel::obtain_from_generic(CKernel* kernel)
+{
+	// TODO: Add proper periodic kernel enum value
+	if (kernel->get_kernel_type()!= K_PERIODIC)
+	{
+		SG_SERROR("CPeriodicKernel::obtain_from_generic(): provided kernel is "
+				"not of type CPeriodicKernel!\n");
+	}
+
+	/* since an additional reference is returned */
+	SG_REF(kernel);
+	return (CPeriodicKernel*)kernel;
+}
+
+void CPeriodicKernel::cleanup()
+{
+	if (sq_lhs != sq_rhs)
+		SG_FREE(sq_rhs);
+	sq_rhs = NULL;
+
+	SG_FREE(sq_lhs);
+	sq_lhs = NULL;
+
+	CKernel::cleanup();
+}
+
+void CPeriodicKernel::precompute_squared_helper(float64_t* &buf, CDotFeatures* df)
+{
+	ASSERT(df)
+	int32_t num_vec=df->get_num_vectors();
+	buf=SG_MALLOC(float64_t, num_vec);
+
+	for (int32_t i=0; i<num_vec; i++)
+		buf[i]=df->dot(i,df, i);
+}
+
+bool CPeriodicKernel::init(CFeatures* l, CFeatures* r)
+{
+	///free sq_{r,l}hs first
+	cleanup();
+
+	CDotKernel::init(l, r);
+	precompute_squared();
+	return init_normalizer();
+}
+
+float64_t CPeriodicKernel::compute(int32_t idx_a, int32_t idx_b)
+{
+	/* Periodic kernel defined as by David Duvenaud in 
+	 http://mlg.eng.cam.ac.uk/duvenaud/cookbook/index.html 
+
+	 k({\bf x},{\bf x'})= exp(-\frac{2sin^2(|{\bf x}-{\bf x'}|/p)}{l^2})
+	*/
+	return exp(-2*pow(sin(M_PI*distance(idx_a,idx_b)/period), 2)/pow(width,2));
+}
+
+void CPeriodicKernel::precompute_squared()
+{
+	if (!lhs || !rhs)
+		return;
+
+	precompute_squared_helper(sq_lhs, (CDotFeatures*) lhs);
+
+	if (lhs==rhs)
+		sq_rhs=sq_lhs;
+	else
+		precompute_squared_helper(sq_rhs, (CDotFeatures*) rhs);
+}
+
+SGMatrix<float64_t> CPeriodicKernel::get_parameter_gradient(
+		const TParameter* param, index_t index)
+{
+	REQUIRE(lhs && rhs, "Features not set!\n")
+
+	if (!strcmp(param->m_name, "width"))
+	{
+		SGMatrix<float64_t> derivative=SGMatrix<float64_t>(num_lhs, num_rhs);
+
+		for (int j=0; j<num_lhs; j++)
+			for (int k=0; k<num_rhs; k++)
+			{
+				float64_t dist=distance(j,k);
+				float64_t trig_arg = M_PI*dist/period;
+				float64_t original = compute(j, k);
+
+				derivative(j,k)=original*4.0*pow(sin(trig_arg),2)/pow(width,3);
+			}
+
+		return derivative;
+	}
+	else if (!strcmp(param->m_name, "period")) 
+	{
+		SGMatrix<float64_t> derivative=SGMatrix<float64_t>(num_lhs, num_rhs);
+
+		for (int j=0; j<num_lhs; j++)
+			for (int k=0; k<num_rhs; k++)
+			{
+				float64_t dist=distance(j,k);
+				float64_t trig_arg = M_PI*dist/period;
+				float64_t original = compute(j, k);
+
+				derivative(j,k)=original*4.0*M_PI*dist*cos(trig_arg)*sin(trig_arg)/pow(period*width,2);
+			}
+		return derivative;
+	}
+	else
+	{
+		SG_ERROR("Can't compute derivative wrt %s parameter\n", param->m_name);
+		return SGMatrix<float64_t>();
+	}
+}
+
+void CPeriodicKernel::init()
+{
+	set_width(1.0);
+	set_period(1.0);
+	sq_lhs=NULL;
```

Comment:
Can you explain why they should be registered? Registered parameters are for model selection or am I misunderstanding something?